be/src/olap/tablet.cpp

```diff
@@ -675,6 +676,62 @@ OLAPStatus Tablet::capture_rs_readers(const std::vector<Version>& version_path,
     return OLAP_SUCCESS;
 }
 
+OLAPStatus Tablet::capture_rs_readers_by_range(const Version& spec_version,
+                                               vector<RowsetReaderSharedPtr>* rs_readers,
+                                               const std::vector<OlapScanRange*>& key_ranges,
+                                               std::shared_ptr<MemTracker> parent_tracker) {
+    std::set<RowsetId> rowsets;
+    bool has_segment_range = false;
+    for (auto key_range : key_ranges) {
+        if (key_range->range_type == ScanRangeType::SEGMENT) {
+            for (const auto& s : key_range->segments) {
+                rowsets.emplace(s.first);
+            }
+            has_segment_range = true;
+        }
+    }
+    if (!has_segment_range) {
```

Comment:
Is it possible to have both KEY and SEGMENT key range in `key_ranges`?
If yes, we should clear the `rs_readers` before calling `capture_rs_readers()`

Reply:
if key_ranges  have SEGMENT key range, it will have KEY key range